Add documentation for telemetry v1 vs v2 (#7939)
istio/istio.io: content/en/docs/ops/configuration/telemetry/telemetry-v1-vs-v2/index.md

---

> title: Differences between Telemetry v1 and v2

Author: Do we actually use the term in-proxy? I have not seen/heard it before but if that's the term to use I can change it.

> title: Differences between Telemetry v1 and v2

Reviewer: I don't know that we have an official term, which is a problem I think. I've been trying to use "in-proxy" or "proxy-based" in my comms, but we need a better term imho.

description: Tradeoffs and gaps between Telemetry v1 and v2.

> weight: 10

Reviewer 2: Who is the audience for this doc?
It reads more like a blog post , i think we should turn it into a blog post.

We can have a paragraph about specific differences / lack of parity on the current telemetry page.

Author: Agree, I tried making it not feel like a blog post but it was very difficult not to. I can convert this to a blog if that's more appropriate.

Author: Issue is there's no good place to add this information in current docs as the surrounding context is not there. After I added the context, it started to look like a blog post.

Reviewer: I don't think we should put specific differences on our current telemetry page. New users won't care -- and it could be more of a distraction than a help.  We'll need somewhere to document the gRPC shim -- and the two should probably be closely-aligned.

owner: istio/wg-policies-and-telemetry-maintainers
test: no
---

> 

Reviewer: Can we provide a summary statement here?  Something like: `Istio has transitioned to a new platform for extensibility and telemetry generation. While this new platform unlocks several critical use cases and provides numerous performance benefits, there are some features of the prior platform that are not supported as of Istio 1.7. The Istio community is working hard to close the remaining gaps, which are discussed below.` ?

Providing uniform telemetry and extension mechanisms to extend the functionality
in Istio have been the driving factors for the architectural choices made in
Istio. From early days, mixer aka Telemetry v1 provided both of these

> functionalities where an out-of-process service gathered attributes from the

Reviewer 3: ```suggestion
functionalities, where an out-of-process service gathered attributes from the
```

> workload Envoy proxies, combined it with platform data like Kubernetes services

Reviewer 3: ```suggestion
workload Envoy proxies and combined it with platform data, like Kubernetes services
```

> and deployment associations and generated rich telemetry data like Prometheus

Reviewer 3: ```suggestion
and deployment associations, and generated rich telemetry data like Prometheus
```

I made some suggestions, but this sentence has a lot of information that may make it hard for people to understand. I'd recommend breaking it up into shorter, sequential sentences.

> metrics, tracing and logs. It also provided an extension mechanism where

Reviewer 3: ```suggestion
metrics, tracing and logs. This model also provided an extension mechanism where
```

> users could write adapters to ingest these attributes and enforce policies or

Reviewer 3: ```suggestion
users could write adapters to ingest these attributes, enforce policies or
```

transform them before exporting it to a third-party monitoring tool.

While this provided a flexible architecture, mixer consumed significant resources

> in clusters with large amounts of workloads and often became a single point of failure.

Reviewer 3: A link to the set of issues within the text would greatly re-enforce this idea. I'm sure there were plenty of other issues fixed.

https://github.com/istio/istio/issues/11400
https://istio.io/latest/blog/2019/performance-best-practices/#istio-1-2-performance

Additionally, the remote call made by the mixer client at Envoy proxies
attributed to performance degradation and telemetry loss. In Istio 1.3 and

> above, mixer-less aka telemetry v2 support was added to address these concerns

Reviewer 3: ```suggestion
above, mixer-less telemetry, aka telemetry v2, support was added to address these concerns
```

while surfacing similar telemetry and extension capabilities as telemetry v1.

> Telemetry v2 works by adding WASM extensions in Envoy proxies for surfacing

Reviewer 4: nitpicks: WASM is spelled Wasm. mixer is spelled Mixer.

Prometheus metrics, thereby removing the out-of-process call to a centralized

> mixer component. Additionally, custom WASM filters can be used to provide mixer

Reviewer: ```suggestion
Telemetry v2 works by adding WASM extensions in Envoy proxies for request and response processing, thereby removing the out-of-process call to a centralized Mixer component. A Prometheus stats extension and a workload metadata exchange extensions are provided out-of-the-box. Additionally, custom WASM filters can be used to provide Mixer
```

adapter like functionality. You can read more about this in the announcement
[blog](/blog/2020/wasm-announce/).

> Moving to telemetry v2 reduces resource cost and improves proxy performance, and

Reviewer 3: ```suggestion
Moving to telemetry v2 reduces resource costs, as there are fewer pods in your environment, and improves proxy performance, and
```

> Moving to telemetry v2 reduces resource cost and improves proxy performance, and
> is the recommended way to gather telemetry from Istio, leading us to deprecate
> support for telemetry v1 in Istio 1.7 release. However, there are some

Reviewer: ```suggestion
Moving to proxy-based telemetry (v2) reduces resource cost and improves proxy performance.
As of 1.7, it is the recommended way to gather telemetry from Istio, with support for Mixer-based
telemetry generation being fully deprecated in Istio 1.7 release (and removed entirely for 1.8).

However, there are some
```

> support for telemetry v1 in Istio 1.7 release. However, there are some

Reviewer 3: I feel like this should be called out earlier, perhaps at the top, of this document.

Such as,

> Support for mixer telemetry in Istio 1.7 has been deprecated and will be removed in Istio 1.8. This blog details the reasons why users should adopt telemetry v2, aka mixer-less telemetry, if they haven't yet, as well as the reasons why the Istio community has decided this is the supported architecture for observability going forward.

differences or gaps in the telemetry reported by telemetry v2 as compared to
telemetry v1 which are important to understand and are discussed below:

> 

Reviewer: Maybe insert something here as a header that is "Known Gaps" or something?

> 1. Telemetry v2 relies on metadata exchange between Envoy proxies to gather

Reviewer: Here and below, I'd start with a summary point -- and then dive into details.

Maybe: `No metadata generated for out-of-mesh traffic`.

   information like peer workload name, namespace and labels. In telemetry v1
   this functionality was performed by mixer as part of combining attributes with
   the platform data. This metadata exchange in telemetry v2 is performed by the
   Envoy proxies by adding specific HTTP header for HTTP protocol or augmenting
   ALPN protocol for TCP protocol as described [here](/docs/tasks/observability/metrics/tcp-metrics/#understanding-tcp-telemetry-collection).

>    This requires Envoy proxies to be injected at both client & server workloads

Reviewer 3: ```suggestion
   This requires Envoy proxies to be injected at both client & server workloads,
```

   implying that the telemetry reported when one peer is not in the mesh will be
   missing peer attributes like workload name, namespace and labels.
   However, if both peers have proxies injected all the labels mentioned [here](/docs/reference/config/metrics/)
   are available in the generated metrics.

> 1. Since metadata exchange for TCP protocol relies on

Reviewer: maybe: `TCP metadata exchange requires mTLS`

>    [Istio ALPN protocol](/docs/tasks/observability/metrics/tcp-metrics/#understanding-tcp-telemetry-collection).

Reviewer 3: ```suggestion
   [Istio ALPN protocol](/docs/tasks/observability/metrics/tcp-metrics/#understanding-tcp-telemetry-collection)
```

   , mutual TLS (mTLS) needs to be enabled for proxies to exchange metadata
   successfully. This implies that if mTLS is not enabled in your cluster,
   telemetry for TCP protocol will not include peer information like
   workload name, namespace and labels.

> 1. Telemetry v1 supported customizing buckets for histogram type metrics like

Reviewer: Maybe: `No mechanism for configuring custom bucketization for histogram metrics`

   request duration and TCP byte sizes. In telemetry v2, there's no such

>    mechanism available. Additionally, the buckets available for latency metrics

Reviewer 3: Maybe a link to mixer charts in 1.7 and link to what they are now? That may be too technical, but the buckets now make a lot more sense as they are orders of 10, if I recall.

   in telemetry v2 are in milliseconds as compared to seconds in telemetry v1.
   However, more buckets are available by default in telemetry v2 for latency
   metrics at the lower latency levels.

> 1. Telemetry v1 supported metric expiration whereby metrics not seen for a

Reviewer: Maybe: `No support for metrics expiry in-proxy`

> 1. Telemetry v1 supported metric expiration whereby metrics not seen for a

Reviewer 3: I would link to https://github.com/istio/istio/issues/21377

```suggestion
1. Telemetry v1 supported metric expiration, whereby metrics not seen for a
```

   configurable amount of time were de-registered for collection. This is
   useful in scenarios where short-lived jobs surfaced telemetry only for a
   short amount of time, and de-registering the metrics prevented reporting of
   metrics which would no longer change in future reducing network traffic and
   storage for Prometheus.

> 1. WASM extension API and plugins are being actively added in the community but

Reviewer: Maybe: `Extensions ecosystem still developing` ???

   we have not reached feature parity with the adapters available for

>    telemetry v1. If you're currently relying on a telemetry v1 adapter you might

Reviewer 3: Is there a list of what adapters have been migrated, if any? I think stackdriver support works great, but I'm not sure about the rest.

   suffer functionality loss when migrating to telemetry v2 if no replacement is
   available.

> We are working on ensuring telemetry v2 reaches feature parity with v1 and

Reviewer: I think I'd add this same bit somewhere up top for extra emphasis.
